pkg/admissioncontroller/webhooks/admission/auditpolicy/admission.go

```diff
@@ -0,0 +1,294 @@
+		if err := h.decoder.Decode(request, cm); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		if !controllerutil.ContainsFinalizer(cm, shoot.FinalizerName) {
+			return acadmission.Allowed("configmap is not referenced by a Shoot")
+		}
+
+		auditPolicy, err := getAuditPolicy(cm)
+		if err != nil {
+			return admission.Errored(http.StatusUnprocessableEntity, err)
+		}
+
+		oldAuditPolicy, ok := oldCm.Data[auditPolicyConfigMapDataKey]
+		if ok {
+			if oldAuditPolicy == auditPolicy {
+				return acadmission.Allowed("audit policy did not change")
+			}
+		}
+
+		// Validate audit policy semantically
+		auditPolicyObj, schemaVersion, err := decoder.Decode([]byte(auditPolicy), nil, nil)
+		if err != nil {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("failed to decode the provided audit policy err=%v", err))
+		}
+		auditPolicyInternal, ok := auditPolicyObj.(*audit_internal.Policy)
+		if !ok {
+			return admission.Errored(http.StatusInternalServerError, fmt.Errorf("failure to cast to audit Policy type: %v", schemaVersion))
+		}
+		errList := auditvalidation.ValidatePolicy(auditPolicyInternal)
+		if len(errList) != 0 {
+			return admission.Errored(http.StatusUnprocessableEntity, fmt.Errorf("provided invalid audit policy err=%v", errList))
+		}
+
+		// Validate schema version against Shoot Kubernetes version
+		existingShoots := &gardencorev1beta1.ShootList{}
+		if err := h.apiReader.List(ctx, existingShoots, client.InNamespace(request.Namespace)); err != nil {
+			return admission.Errored(http.StatusInternalServerError, err)
+		}
+		if err := validatePolicyAgainstShoots(request, schemaVersion, existingShoots); err != nil {
+			return admission.Errored(http.StatusUnprocessableEntity, err)
+		}
+
+		return acadmission.Allowed("configmap change is valid")
+	}
+	return acadmission.Allowed("operation is not update or delete")
+}
+
+// IsValidAuditPolicyVersion checks whether the api server support the provided audit policy apiVersion
+func IsValidAuditPolicyVersion(shootVersion string, schemaVersion *schema.GroupVersionKind) (bool, error) {
+	auditGroupVersion := schemaVersion.GroupVersion().String()
+
+	if auditGroupVersion == "audit.k8s.io/v1" {
+		return version.CheckVersionMeetsConstraint(shootVersion, ">= v1.12")
+	}
+	return true, nil
+}
+
+func (h *handler) getOldObject(request admission.Request, oldObj runtime.Object) error {
+	if len(request.OldObject.Raw) != 0 {
+		if err := h.decoder.DecodeRaw(request.OldObject, oldObj); err != nil {
+			return err
+		}
+	} else {
+		return fmt.Errorf("could not find old object")
+	}
+	return nil
+}
+
+func getAuditPolicy(cm *corev1.ConfigMap) (auditPolicy string, err error) {
+	auditPolicy, ok := cm.Data[auditPolicyConfigMapDataKey]
+	if !ok {
+		return "", fmt.Errorf("missing '.data.policy' in audit policy configmap")
+	}
+	if len(auditPolicy) == 0 {
+		return "", fmt.Errorf("empty audit policy. Provide non-empty audit policy")
```

Comment:
Empty audit policy is valid audit policy. We are currently using it, ref:
- https://github.com/gardener/gardener/blob/b84868a76c38343ec3add63055320e12878418e7/charts/seed-controlplane/charts/kube-apiserver/values.yaml#L112
- https://github.com/gardener/gardener/blob/b84868a76c38343ec3add63055320e12878418e7/charts/seed-controlplane/charts/kube-apiserver/templates/configmap-audit-policy.yaml#L9

Reply:
There was the same check before:
https://github.com/gardener/gardener/pull/3720/files#diff-201c3f0c956474991f7e0ebb09d5cb2c23b1b8dbfdad19c36c81c188ce1f13a5L862
How did we deploy an empty policy?

Reply:
Excuse me for the inconvenience, it is my mistake.
It is not empty policy, but a policy with a single rule disabling audit logging https://github.com/gardener/gardener/blob/4c6153dceae2e461607c9a38afca88277387dda0/charts/seed-controlplane/charts/kube-apiserver/templates/_helpers.tpl#L93-L96.